server/src/main/java/io/druid/server/coordinator/CostBalancerStrategy.java

```diff
@@ -316,30 +327,64 @@ protected double computeCost(
    * For assignment, we want to move to the lowest cost server that isn't already serving the segment.
    *
    * @param proposalSegment A DataSegment that we are proposing to move.
-   * @param serverHolders   An iterable of ServerHolders for a particular tier.
+   * @param serverHolders   A list of ServerHolders for a particular tier.
    *
    * @return A ServerHolder with the new home for a segment.
    */
 
   protected Pair<Double, ServerHolder> chooseBestServer(
       final DataSegment proposalSegment,
-      final Iterable<ServerHolder> serverHolders,
+      final List<ServerHolder> serverHolders,
       final boolean includeCurrentServer
   )
   {
+    final long proposalSegmentSize = proposalSegment.getSize();
+    final List<ServerHolder> filtered = Lists.newArrayList(
+        Iterables.filter(
+            serverHolders, new Predicate<ServerHolder>()
+            {
+              @Override
+              public boolean apply(ServerHolder server)
+              {
+                // (optional) Don't include server if it is already serving segment
+                if (!includeCurrentServer && server.isServingSegment(proposalSegment)) {
+                  return false;
+                }
+                // Don't calculate cost if the server doesn't have enough space or is loading the segment
+                if (proposalSegmentSize > server.getAvailableSize() || server.isLoadingSegment(proposalSegment)) {
+                  return false;
+                }
+                return true;
+              }
+            }
+        )
+    );
+
     Pair<Double, ServerHolder> bestServer = Pair.of(Double.POSITIVE_INFINITY, null);
 
+    int numServer = filtered.size();
+    int size = numServer / threadCount + (numServer % threadCount == 0 ? 0 : 1);
+
     List<ListenableFuture<Pair<Double, ServerHolder>>> futures = Lists.newArrayList();
 
-    for (final ServerHolder server : serverHolders) {
+    for (final List<ServerHolder> servers : Lists.partition(filtered, size)) {
```

Comment:
Total amount of calculation would be the least if all is done by single thread because minimum cost is calculated per thread. But it can be slower than multi-threaded calculation without early-escape. So it's rough work partitioning based on size of thread-pool. I tried shared variable but seemed not good.